rosbag2_transport/src/rosbag2_transport/rosbag2_transport.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rosbag2_transport/rosbag2_transport.hpp"
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <queue>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcl/graph.h"
+#include "rclcpp/rclcpp.hpp"
+#include "rcutils/time.h"
+
+#include "rosbag2_transport/logging.hpp"
+#include "rosbag2/sequential_reader.hpp"
+#include "rosbag2/types.hpp"
+#include "rosbag2/typesupport_helpers.hpp"
+#include "rosbag2/writer.hpp"
+
+#include "generic_subscription.hpp"
+#include "player.hpp"
+#include "replayable_message.hpp"
+#include "rosbag2_node.hpp"
+
+namespace rosbag2_transport
+{
+
+Rosbag2Transport::Rosbag2Transport()
+: reader_(std::make_shared<rosbag2::SequentialReader>()),
+  writer_(std::make_shared<rosbag2::Writer>())
+{}
+
+Rosbag2Transport::Rosbag2Transport(
+  std::shared_ptr<rosbag2::SequentialReader> reader, std::shared_ptr<rosbag2::Writer> writer)
+: reader_(std::move(reader)), writer_(std::move(writer)) {}
+
+void Rosbag2Transport::init()
+{
+  rclcpp::init(0, nullptr);
+}
+
+void Rosbag2Transport::shutdown()
+{
+  rclcpp::shutdown();
+}
+
+void Rosbag2Transport::record(
+  const StorageOptions & storage_options, const RecordOptions & record_options)
+{
+  writer_->open(storage_options);
+
+  auto rosbag2_transport = std::make_shared<Rosbag2Node>("rosbag2");
```

Comment:
does it make sense to have one node per rosbag transport? If I want to subscribe and record at the same time, I have twice the same node set up, right?

Reply:
Sounds sensible to me. That way we can also make sure that no unnecessary nodes are created and we always work with exactly one node.

Reply:
I implemented a simple lazy-initialization cache in `rosbag2_transport` which will do the trick. We can't inject the node because (for testing) we need to control when and if we call `rclcpp::init(...)`.